### PR TITLE
PP-4976 Implement form submission for organisation address

### PR DESCRIPTION
**Pull request by stephencdaly**
Send the organisation details and update the go live stage as a single request when the form is submitted and validates successfully.

Handle the ENTERED_ORGANISATION_ADDRESS go-live stage on the go live index page and for displaying the link on the dashboard.

The organisation address page is not currently part of the flow, but behaves correctly in isolation.

### CODE REVIEW

File: app/controllers/request-to-go-live/organisation-address/post.controller.js [hunk 1]
```diff
@@ -2,44 +2,51 @@
...
 const { renderErrorView } = require('../../../utils/response')
+const { validPaths, ServiceUpdateRequest } = require('../../../models/ServiceUpdateRequest.class')
 
-const ADDRESS_LINE1_FIELD = 'address-line1'
-const ADDRESS_LINE2_FIELD = 'address-line2'
-const ADDRESS_CITY_FIELD = 'address-city'
-const ADDRESS_COUNTRY_FIELD = 'address-country'
-const ADDRESS_POSTCODE_FIELD = 'address-postcode'
-const TELEPHONE_NUMBER_FIELD = 'telephone-number'
+const clientFormIds = {
```
Comment (sfount) [inline app/controllers/request-to-go-live/organisation-address/post.controller.js]: 🥇

File: test/unit/clients/adminusers_client/service/update_service_test.js [hunk 2]
```diff
@@ -139,28 +133,21 @@ describe('adminusers client - patch request to update service', function () {
       'value': 'bar'
     }]
 
-    before(done => {
+    before(() => {
       provider.addInteraction(
```
Comment (sfount) [inline test/unit/clients/adminusers_client/service/update_service_test.js]: If you are removing `done` you should return the promise back up the chain
```suggestion
      return provider.addInteraction(
```

File: test/unit/clients/adminusers_client/service/update_service_test.js [hunk 3]
```diff
@@ -98,14 +95,16 @@ describe('adminusers client - patch request to update service', function () {
...
     const validUpdateServiceResponse = serviceFixtures.validServiceResponse({
       external_id: existingServiceExternalId,
-      merchant_details: merchantDetails
+      merchant_details: merchantDetails,
+      current_go_live_stage: currentGoLiveStage
     })
 
-    before(done => {
+    before(() => {
       provider.addInteraction(
```
Comment (sfount) [inline test/unit/clients/adminusers_client/service/update_service_test.js]:
```suggestion
      return provider.addInteraction(
```

File: test/unit/clients/adminusers_client/service/update_service_test.js [hunk 4]
```diff
@@ -114,21 +113,16 @@ describe('adminusers client - patch request to update service', function () {
...
     afterEach(() => provider.verify())
 
-    it('should update service successfully', (done) => {
-      adminUsersClient.updateService(existingServiceExternalId, validUpdateServiceRequest)
-        .should.be.fulfilled
-        .then(service => {
-          expect(service.externalId).to.equal(existingServiceExternalId)
-          expect(service.merchantDetails).to.deep.equal(merchantDetails)
-        }).should.notify(done)
+    it('should update service successfully', async function () {
```
Comment (sfount) [inline test/unit/clients/adminusers_client/service/update_service_test.js]: 🥇

File: test/unit/clients/adminusers_client/service/update_service_test.js [hunk 5]
```diff
@@ -50,7 +51,7 @@ describe('adminusers client - patch request to update service', function () {
       }
     })
 
-    before(done => {
+    before(() => {
       provider.addInteraction(
```
Comment (sfount) [inline test/unit/clients/adminusers_client/service/update_service_test.js]:
```suggestion
      return provider.addInteraction(
```

File: app/controllers/request-to-go-live/organisation-address/post.controller.test.js [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+        await controller(req, res)
+
+        expect(mockServiceService.updateService.calledWith(serviceExternalId, expectedUpdateServiceRequest, correlationId)).to.equal(true)
+        expect(res.redirect.calledWith(303, `/service/${serviceExternalId}/request-to-go-live/choose-how-to-process-payments`)).to.equal(true)
+      })
+    })
+
+    describe('error updating service', done => {
+      it('should show an error page if updating service throws error', done => {
```
Comment (sfount) [inline app/controllers/request-to-go-live/organisation-address/post.controller.test.js]: 👍 Testing error cases

File: app/controllers/request-to-go-live/organisation-address/post.controller.js [hunk 7]
```diff
@@ -82,19 +89,42 @@ module.exports = (req, res) => {
...
+      telephone_number: formFields[clientFormIds.telephoneNumber]
     })
-    return res.redirect(303, requestToGoLive.organisationAddress.replace(':externalServiceId', req.service.externalId))
+    res.redirect(303, requestToGoLive.organisationAddress.replace(':externalServiceId', req.service.externalId))
+  }
+}
+
+module.exports = async function (req, res) {
+  try {
+    await submit(req, res)
```
Comment (sfount) [inline app/controllers/request-to-go-live/organisation-address/post.controller.js]: I know this is a pattern we have in a number of places but I'm generally uncomfortable with passing the entire `req` and `res` object around.

I don't have the full context for this feature but could this be split into something like: 
```js 
try { 
  const { errors, form } = validateFormBody(req.body) 
  
  if (lodash.isEmpty(errors) { 
    // use form values - build service update request and `await` submit
  } else { 
    // use error values - render error view
  }
} catch (error) { 
  logger.error(`Error submitting organisation address - ${err.stack}`)
  renderErrorView(req, res)
}
```

Comment (stephencdaly) [reply app/controllers/request-to-go-live/organisation-address/post.controller.js]: nice suggestion, i'll do it like this

File: app/controllers/request-to-go-live/organisation-address/post.controller.js [hunk 8]
```diff
@@ -2,99 +2,136 @@
...
-    ADDRESS_CITY_FIELD,
-    ADDRESS_COUNTRY_FIELD,
-    ADDRESS_POSTCODE_FIELD,
-    TELEPHONE_NUMBER_FIELD
-  ]
-  const formFields = fields.reduce((form, field) => {
-    form[field] = trimField(field, req.body)
-    return form
-  }, {})
+const submitForm = async function (form, serviceExternalId, correlationId) {
```
Comment (sfount) [inline app/controllers/request-to-go-live/organisation-address/post.controller.js]: This is a really clean design, nice! Splits apart logic and deals with specific parts of the request (bonus for handling Promises well)

**Pull request merged**